executor/runcexecutor/executor.go

```diff
@@ -310,6 +310,7 @@ func (w *runcExecutor) Run(ctx context.Context, id string, root executor.Mount,
 				timeout()
 				select {
 				case <-time.After(50 * time.Millisecond):
+					cancelRun()
```

Comment:
This allows the tests to pass, I am a little confused why this was necessary.  The pid1 `sleep 10` was continuing to run, even though the `ctx.Done` was triggered almost immediately.  Then the `w.runc.Kill` ran which did not return an error, but the `w.run` call below didn't terminate.   So this loop waited for 50ms then ran the `kill 9` again and again, until the pid1 ended after 10s.  I am not sure how the `kill 9` is not actually terminating the runc process.  With this change we cancel the ctx passed into `w.run` after the `kill 9` is ignored, and the process ends up exiting as expected.

Reply:
This does not look quite right. I'm not sure why sigkill wouldn't work (can we replicate it outside of test?) but even when we handle misbehaving runc that doesn't react(properly) to sigkill, we should give it more than 50ms to shut down normally.

Reply:
I have tried to reproduce the steps by calling runc directly and so far it works, not sure what the difference is.  I will keep looking, although I have very little time these days.  The only thing I can think of is that the sigkill signal is not actually being sent to the runc or sleep process, not sure how/why though.

Reply:
if you can see that this work outside of the test. can you just add a counter before calling `cancelRun()` so that case does not get called on 50ms but after a couple of seconds for example? Also some comments describing why we are doing this.

Reply:
I spent some time debugging this, I think the problem is with the test container, not with buildkit.  The sleep process gets wedged on `zap_pid_ns_processes` which seems to be related to parent/child reaping.  The `cancelRun` was just giving up and shutting things down even though the sleep process persisted. I have updated the test container to run with `tini` as the entrypoint to better handle reaping during the go tests, so far it seems to work well.  Hopefully it will make it through the github workflows...

Reply:
Looks like the tests all pass when run under `tini` with out the `cancelRun()` hack.

Reply:
I have another idea why the zombie processes might happening, let me test that out.  In theory runc should be doing the waitpid so there should be no zombies for `tini` to handle...

Reply:
I have updated the PR again, the bug was in `(*runcExecutor).Exec`, and the `tini` hack put me on the right path.  So the runc pid1 (sleep) was getting wedged on `zap_pid_ns_process`, which according to the docs:
> The `zap_pid_ns_processes` function is used in Linux to terminate all processes within a specific namespace

So the problem was not with the pid1 Kill directly, it was actually a zombie process from the `Exec` (the `sh` command) which was preventing `zap_pid_ns_processes` from finishing.  It turns out we were using `context.Background()` for the `runc` call via `(*runcExecutor).Run` but were using the request `ctx` for the `runc` call via `(*runcExecutor).Exec`.  So when the `ctx.Done` happened on the parent `runc` command for the `Exec` was getting terminated immediately, before it could call waitpid on the child  `sh` command.

I have moved the wait/kill logic into a common routine now that will be called for both `Run` and `Exec` and we use `context.Background()` for both `runc` calls now.